edb/edgeql/parser/grammar/expressions.py

```diff
@@ -907,80 +906,16 @@ def reduce_NFCONST(self, *kids):
         self.val = qlast.DecimalConstant(value=kids[0].val)
 
 
-class RawStringConstant(Nonterm):
-
-    def reduce_RSCONST(self, str_tok):
-        match = lexutils.VALID_RAW_STRING_RE.match(str_tok.val)
-        if not match:
-            raise EdgeQLSyntaxError(
-                f"invalid raw string literal", context=str_tok.context)
-
-        quote = match.group('Q')
-        val = match.group('body')
-
-        self.val = qlast.RawStringConstant(value=val, quote=quote)
-
-
-class EscapedStringConstant(Nonterm):
-
-    def reduce_SCONST(self, str_tok):
-        match = lexutils.VALID_STRING_RE.match(str_tok.val)
-
-        if not match:
-            raise EdgeQLSyntaxError(
-                f"invalid string literal", context=str_tok.context)
-        if match.group('err_esc'):
-            raise EdgeQLSyntaxError(
-                f"invalid string literal: invalid escape sequence "
-                f"'{match.group('err_esc')}'",
-                context=str_tok.context)
-        elif match.group('err_cont'):
-            raise EdgeQLSyntaxError(
-                f"invalid string literal: invalid line continuation",
-                hint="newline has to immediately follow '\\'",
-                context=str_tok.context)
-
-        quote = match.group('Q')
-        val = match.group('body')
-
-        # collapse the whitespace after a line continuation
-        val = lexutils.collapse_newline_whitespace(val)
```

Comment:
The current lexer has a nice property: it's mostly non-destructive.  We can reconstruct the input stream by concatenating the tokens back. This property might be important in some contexts.  @vpetrovykh should know more.

Reply:
Technically there is `token.text()` which keeps everything and `token.value()`, which extracts semantically valid value. But I think we can get rid of `token.text()` in the future to avoid allocations. I don't see the reason to keep it.

Reply:
Or rather `.value` and `.clean_value` in pythonic tokens

Reply:
Well, that was about the token level. The AST is "abstract". You can't reconstruct exact text anymore (you're skipping comments and whitespace). So there is no reason to pick some single thing that we can reconstruct: quotes. *(there is even a concept called CST that keeps everything)*

Reply:
Yes, I didn't mean reconstructing from AST, only that the lexer has this capability:

> We can reconstruct the input stream by concatenating the tokens back.